content/en/blog/_posts/2022-11-11-retroactive-default-storage-class.md

```diff
@@ -0,0 +1,131 @@
+
+## What changed
+
+We've changed PersistentVolume (PV) controller in KCM, so it can assign default StorageClass to unbound PersistentVolumeClaims that have storageClassName set to `nil`. In order to allow the value change PersistentVolumeClaim admission in the API server also changed to allow changing the value from `nil` to actual StorageClass name.
+
+### `nil` vs `""` for storageClassName - does it matter?
+
+Before this feature was introduced those values were equal in terms of behavior and any PersistentVolumeClaim with `nil` or `""` storageClassName value would bind to existing PersistentVolume resource with storageClassName also set to `nil` or `""`.
+
+With this new feature enabled we wanted to maintain this behavior but also be able to update the StorageClass name. With these constraints in mind the feature changes the semantics of `nil` if default StorageClass is present to always mean "Give me a default" and `""` to mean "Give me PersistentVolume that also has `""` StorageClass name".
+
+In other words we need to distinguish two main cases, either default StorageClass is not present in which case there's no behavior change or where there is a default in which case PVCs with `nil` won't bind to PVs anymore and instead will be updated to have the StorageClass.
+
+The tables below show all these cases to better describe when PVC binds and when it's StorageClass gets updated.
+
+| Without default class        | PVC storageClassName == `""` | PVC storageClassName == `nil` |
+|------------------------------|------------------------------|-------------------------------|
+| PV storageClassName == `""`  | binds                        | binds                         |
+| PV without storageClassName  | binds                        | binds                         |
+
+
+| With default class          | PVC storageClassName == `""` | PVC storageClassName == `nil` |
+|-----------------------------|------------------------------|-------------------------------|
+| PV storageClassName == `""` | binds                        | class updates                 |
+| PV without storageClassName | binds                        | class updates                 |
+
+## How to use it
+
+You don't have to perform any configuration steps - the feature will work "out of the box" once it's GA.
+
+If you want to test the feature in Alpha you need to enable FeatureGates in `kube-controller-manager` and `kube-apiserver`:
+
+```bash
+--feature-gates="...,RetroactiveDefaultStorageClass=true"
+```
+
+### Test drive
+
+If you would like to see the feature in action and verify it works fine in your cluster here's what you can try:
+
+1. We'll use a very simple PersistentVolumeClaim.
+```yaml
+apiVersion: v1
+kind: PersistentVolumeClaim
+metadata:
+  name: csi-pvc
```

Comment:
How do we know this will use a CSI driver?

Reply:
We don't in fact. I just had a cluster with only one csi driver present but surely it makes more sense to use a generic name here, changed it.